txstratum/middleware.py

```diff
@@ -0,0 +1,107 @@
+            # Then check if version is the minimum allowed
+            # Example of a wallet desktop user agent
+            # Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) HathorWallet/0.22.1
+            # Chrome/73.0.3683.121 Electron/5.0.13 Safari/537.36 HathorWallet/0.22.1
+            wallet_desktop_regex = r"HathorWallet/(\d+\.\d+.\d+)"
+            if is_version_invalid(
+                wallet_desktop_regex, user_agent, min_wallet_desktop_version
+            ):
+                return web.json_response(
+                    {"error": VERSION_CHECK_ERROR_MESSAGE}, status=400
+                )
+
+        if min_wallet_mobile_version:
+            # Search user agent for wallet mobile string and get version
+            # Then check if version is the minimum allowed
+            # Example of a wallet mobile user agent
+            # Hathor Wallet Mobile / 0.18.0
+            wallet_mobile_regex = r"Hathor Wallet Mobile / (\d+\.\d+\.\d+)"
+            if is_version_invalid(
+                wallet_mobile_regex, user_agent, min_wallet_mobile_version
+            ):
+                return web.json_response(
+                    {"error": VERSION_CHECK_ERROR_MESSAGE}, status=400
+                )
+
+            # Before version 0.18.0 in the wallet mobile, the user agent was receiving a fixed "version"
+            # the user agent was "HathorMobile/1", so if the min version is at least 0.18.0, then
+            # we must have a custom check here for it
+            if is_version_gte(min_wallet_mobile_version, "0.18.0"):
+                custom_regex = "HathorMobile/1"
+                search = re.search(custom_regex, user_agent)
+                if search:
+                    # If the regex found it, then we should block
+                    return web.json_response(
+                        {"error": VERSION_CHECK_ERROR_MESSAGE}, status=400
+                    )
+
+        if min_wallet_headless_version:
+            # Seach user agent for wallet headless string and get version
+            # Then check if version is the minimum allowed
+            # Example of a wallet headless user agent
+            # Hathor Wallet Headless / 0.14.0
+            wallet_headless_regex = r"Hathor Wallet Headless / (\d+\.\d+\.\d+)"
+            if is_version_invalid(
+                wallet_headless_regex, user_agent, min_wallet_headless_version
+            ):
+                return web.json_response(
+                    {"error": VERSION_CHECK_ERROR_MESSAGE}, status=400
+                )
+
+        response = await handler(request)
+        return response
+
+    return version_check
+
+
+def is_version_invalid(regex: str, user_agent: str, min_version: str) -> bool:
+    """Check if version in user agent is invalid comparing with min version.
+
+    Regex is used to get the version from user agent
+    """
+    search = re.search(regex, user_agent)
```

Comment:
It is possible that compiling the regex beforehand (outside this function) would be more efficient. By using `re.search` every time, we could be compiling then every time as well.

https://docs.python.org/3/library/re.html#re.compile

Although for this case that has just a few regex'es I'm not really sure it makes any difference, because apparently the `re` module caches some of the compiled regex'es.

But I would say we should compile them explicitly even if there is no gain right now, to make it explicit that we are trying to optimize, and because we don't know how many more regex'es we could add in the future

Reply:
I can do it but I feel wouldn't help much to be honest, each `re.search` I do is for a different regex, so I can't reuse them. I think this would be helpful if we were running many searches for a single regex, which is not the case.

Anyway, I can do it no problem. Do you feel it's worth it compiling now even with different regex to prevent loss of performance in the future?